[JIT] Allow module interface subtyping to skip argument name matching

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47950 [JIT] Allow module interface subtyping to skip argument name matching**

**Summary**
This commit allows users to define a class attribute in a module
interface definition that specifies which arguments of which methods
should be ignored when determining if modules subtype the interface in
question. Methods of interfaces that allow name mismatches cannot be
called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

Differential Revision: [D24966648](https://our.internmc.facebook.com/intern/diff/D24966648)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @cbalioglu @gcramer23

Commit:
[WIP][JIT] Allow module interface subtyping to skip argument name matching

**Summary**
This commit allows `@torch.jit.interface` to accept a boolean argument
that specify whether implementations must have the same argument names
for matching methods. Methods of interfaces that allow name mismatches cannot
be called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

[ghstack-poisoned]

SplitInfinity commented:
**Pros**
👍  simple, restricted to interfaces only
👍  no need to introduce positional args which only exist in 3.7+

**Cons**
👎  BC-breaking (but the fix is a simple find and replace!)

Commit:
Update on "[WIP][JIT] Allow module interface subtyping to skip argument name matching"


**Summary**
This commit allows `@torch.jit.interface` to accept a boolean argument
that specify whether implementations must have the same argument names
for matching methods. Methods of interfaces that allow name mismatches cannot
be called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

Differential Revision: [D24966648](https://our.internmc.facebook.com/intern/diff/D24966648)

[ghstack-poisoned]

eellison commented:
I haven't looked into the details of this, but my preference would be for the implementation to set "ignore argument name" to be an argument-level property, instead of a property of the whole schema. The former sets us up to easily support the 3.7 positional-only arg syntax, and is arguably a better API for users of 3.7. It's also "a thing" in python, whereas whole function schema no args isn't.

Other comment is that it's not backwards-incompatible if we add a default - 

def `interface(*, ignore_argument_names=False)` (keyword only is debatable but IMO it's a little more explicit and readable)

SplitInfinity commented:
What do you think of using a class attribute to indicate which arguments should be ignored?

eellison commented:
Yea, I could see something like:
```
@torch.jit.interface
def module():
    ___ignored_argument_names = {"forward": ['x', 'y'], "other_method": ...}

```

eellison commented:
The API though, we can do whatever. More just that I htink it would be nice from an implementation perspective to make it easy to switch to the 3.7 syntax when that comes.

Commit:
Update on "[WIP][JIT] Allow module interface subtyping to skip argument name matching"


**Summary**
This commit allows `@torch.jit.interface` to accept a boolean argument
that specify whether implementations must have the same argument names
for matching methods. Methods of interfaces that allow name mismatches cannot
be called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

Differential Revision: [D24966648](https://our.internmc.facebook.com/intern/diff/D24966648)

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow module interface subtyping to skip argument name matching"


**Summary**
This commit allows users to define a class attribute in a module
interface definition that specify which arguments of which methods
should be ignored when determining if modules subtype the interface in
question. Methods of interfaces that allow name mismatches cannot be
called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

Differential Revision: [D24966648](https://our.internmc.facebook.com/intern/diff/D24966648)

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow module interface subtyping to skip argument name matching"


**Summary**
This commit allows users to define a class attribute in a module
interface definition that specifies which arguments of which methods
should be ignored when determining if modules subtype the interface in
question. Methods of interfaces that allow name mismatches cannot be
called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

Differential Revision: [D24966648](https://our.internmc.facebook.com/intern/diff/D24966648)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -264,11 +326,29 @@ struct SourceImporterImpl : public Resolver,
     } else if (superclass_name == "Interface") {
+      // Parse the set of ignored arguments for each method from the ClassDef.
+      auto ignored_argument_names = getIgnoredArgumentNames(class_def);
+      // Strip the Assign for the ignored argument names from the ClassDef for
+      // the interface because define_interface doesn't handle those.
+      ClassDef stripped_class_def = stripNonDefFromClass(class_def);
```
This is not elegant but is a necessary consequence of the fact that `ClassDef` is used to store the AST for interface types and that the parsing machinery for class types is reused for interfaces.

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -752,7 +752,7 @@ def call_prepare_scriptable_func(obj):
                 sub_module[k] = call_prepare_scriptable_func(v)
             obj.__setattr__(name, sub_module)
         elif isinstance(sub_module, torch.nn.Module) and not isinstance(sub_module, ScriptModule):
-            obj.__setattr__(name, call_prepare_scriptable_func(sub_module)) 
```
Oops, need to undo.

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -907,7 +907,7 @@ def forward(self, input):
         return obj
 
     if isinstance(obj, torch.nn.Module):
-        obj = call_prepare_scriptable_func(obj) 
```
Oops, need to undo.

SplitInfinity reviewed (commented):
I left some comments to highlight items I'd like to discuss.

gmagogsfm commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -348,8 +348,9 @@ struct FunctionSchema {
+  // ignored_arg_names - set of argument names to ignore
   // the first argument, which will be the object in both cases
-  bool isSubtypeOf(const FunctionSchema& rhs, bool as_method, std::ostream* why_not=nullptr) const;
+  bool isSubtypeOf(const FunctionSchema& rhs, bool as_method, const std::set<std::string>& ignored_arg_names, std::ostream* why_not=nullptr) const;
```
As in ignoring argument names difference won't be useful at all for first-class functions? In that case, having `ignored_arg_names` still doesn't hurt I think.

gmagogsfm commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -270,14 +270,15 @@ inline FunctionSchema FunctionSchema::cloneWithRemappedTypes(
     const Argument& c = child[i];
     const Argument& p = parent[i];
-    if (c.name() != p.name()) {
+    if ((ignored_arg_names.count(c.name()) == 0) && (c.name() != p.name())) {
```
If I take a step back, we are working on this solution mainly to unblock tracing issue that could generate arbitrary argument names. From a user's perspective, it is difficult to know which argument names would be changed, even harder to know what they are changed to. Allowing specifying either child/parent argument name would help in this case.

SplitInfinity replied:
What's the best way to express this? Boolean argument? Enum argument? It feels very awkward.

gmagogsfm commented on torch/csrc/jit/frontend/sugared_value.h:
```diff
@@ -467,6 +467,15 @@ struct MethodValue : public SugaredValue {
+        if ((ignored_arg_names.find(method_name) != ignored_arg_names.end()) &&
+            !kwargs.empty()) {
+          throw ErrorReport(loc)
+              << method_name << " cannot be called with keyword arguments";
```
Maybe speak a bit more about why kwarg not allowed so user can make the connection between this error and `__ignored_argument_names__ `

SplitInfinity replied:
Done.

gmagogsfm commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1369,11 +1369,30 @@ void initJitScriptBindings(PyObject* module) {
          ResolutionCallback rcb,
-         bool is_module) {
+         bool is_module,
+         py::handle py_ignored_argment_names) {
```
Optional:
You can provide a default value for py_ignored_argument_names so that this change is completely BC, just in case anyone is using this internal method.

Nit:
s/argment/argument

gmagogsfm commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -247,6 +247,68 @@ struct SourceImporterImpl : public Resolver,
+        auto var = Var(assign.lhs());
+
+        // The LHS must be a Var named "__ignored_argument_names__".
+        if (var.name().name() == "__ignored_argument_names__") {
```
Can there be multiple `__ignored_argument_names__`? if not, we can exit this method early after finding one.

SplitInfinity replied:
Added early return.

Commit:
Update on "[JIT] Allow module interface subtyping to skip argument name matching"


**Summary**
This commit allows users to define a class attribute in a module
interface definition that specifies which arguments of which methods
should be ignored when determining if modules subtype the interface in
question. Methods of interfaces that allow name mismatches cannot be
called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

Differential Revision: [D24966648](https://our.internmc.facebook.com/intern/diff/D24966648)

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow module interface subtyping to skip argument name matching"


**Summary**
This commit allows users to define a class attribute in a module
interface definition that specifies which arguments of which methods
should be ignored when determining if modules subtype the interface in
question. Methods of interfaces that allow name mismatches cannot be
called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

Differential Revision: [D24966648](https://our.internmc.facebook.com/intern/diff/D24966648)

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow module interface subtyping to skip argument name matching"


**Summary**
This commit allows users to define a class attribute in a module
interface definition that specifies which arguments of which methods
should be ignored when determining if modules subtype the interface in
question. Methods of interfaces that allow name mismatches cannot be
called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

Differential Revision: [D24966648](https://our.internmc.facebook.com/intern/diff/D24966648)

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow module interface subtyping to skip argument name matching"


**Summary**
This commit allows users to define a class attribute in a module
interface definition that specifies which arguments of which methods
should be ignored when determining if modules subtype the interface in
question. Methods of interfaces that allow name mismatches cannot be
called with named arguments.

**Test Plan**
This commit adds a unit test that tests that implementations with
mismatched arguments subtype properly if the interface allows it.

Differential Revision: [D24966648](https://our.internmc.facebook.com/intern/diff/D24966648)

[ghstack-poisoned]
